### PR TITLE
Add backup selectors for backup-fetch and backup-push

### PR DESCRIPTION
**Pull request by usernamedt**
In this PR I propose to add backup selectors for the additional flexibility of `backup-push` and `backup-fetch` commands.

New functionality for `backup-push` (postgres):
1. Create delta from backup specified by backup name via `--delta-from-name` flag or `WALG_DELTA_FROM_NAME` variable. Sample usage:
```wal-g backup-push /path --delta-from-name base_000000010000000100000072_D_000000010000000100000063```
2. Create delta from backup specified by UserData via `--delta-from-user-data` flag or `WALG_DELTA_FROM_USER_DATA` variable. Sample usage:
```wal-g backup-push /path --delta-from-user-data "{ \"x\": [3], \"y\": 4 }" ```

New functionality for `backup-fetch`:
1. Find the storage backup with the specific UserData via `--target-user-data` flag or `WALG_FETCH_TARGET_USER_DATA` variable and fetch it. Sample usage:
```wal-g backup-fetch /path --target-user-data "{ \"x\": [3], \"y\": 4 }"```

When using the above flags in combination with `WALG_DELTA_ORIGIN` setting, `WALG_DELTA_ORIGIN` logic applies to the specified backup. For example:
```
list of backups in storage:
base_000000010000000100000040  # full backup
base_000000010000000100000046_D_000000010000000100000040  # 1st delta
base_000000010000000100000061_D_000000010000000100000046  # 2nd delta
base_000000010000000100000070  # full backup

export WALG_DELTA_ORIGIN=LATEST_FULL
wal-g backup-push /path --delta-from-name base_000000010000000100000046_D_000000010000000100000040

wal-g logs:
INFO: Selecting the backup with name base_000000010000000100000046_D_000000010000000100000040 as the base for the current delta backup...
INFO: Delta will be made from full backup.
INFO: Delta backup from base_000000010000000100000040 with LSN 140000060.
```

### CODE REVIEW

File: cmd/pg/backup_fetch.go [hunk 1]
```diff
@@ -16,18 +19,26 @@ For information about pattern syntax view: https://golang.org/pkg/path/filepath/
 	restoreSpecDescription        = "Path to file containing tablespace restore specification"
 	reverseDeltaUnpackDescription = "Unpack delta backups in reverse order (beta feature)"
 	skipRedundantTarsDescription  = "Skip tars with no useful data (requires reverse delta unpack)"
+	targetUserDataDescription     = "Fetch storage backup which has the specified UserData"
```
Comment (g0djan) [inline cmd/pg/backup_fetch.go:22]: UserData looks like a label. Let's rename it.

File: cmd/pg/backup_fetch.go [hunk 2]
```diff
@@ -16,18 +19,26 @@ For information about pattern syntax view: https://golang.org/pkg/path/filepath/
...
 
 var fileMask string
 var restoreSpec string
 var reverseDeltaUnpack bool
 var skipRedundantTars bool
+var targetUserData string
 
 var backupFetchCmd = &cobra.Command{
-	Use:   "backup-fetch destination_directory backup_name",
+	Use:   "backup-fetch destination_directory [backup_name OR target-* flag]",
```
Comment (g0djan) [inline cmd/pg/backup_fetch.go:32]: Why not like this ? 
[backup_name] [--target-user-data \<data\>]

File: cmd/pg/backup_fetch.go [hunk 3]
```diff
@@ -16,18 +19,26 @@ For information about pattern syntax view: https://golang.org/pkg/path/filepath/
...
 	Short: backupFetchShortDescription, // TODO : improve description
-	Args:  cobra.ExactArgs(2),
+	Args:  cobra.RangeArgs(1, 2),
 	Run: func(cmd *cobra.Command, args []string) {
+		if targetUserData == "" {
+			targetUserData = viper.GetString(internal.TargetUserDataSetting)
```
Comment (g0djan) [inline cmd/pg/backup_fetch.go:37]: You don't use the variable later.

File: cmd/pg/backup_fetch.go [hunk 4]
```diff
@@ -40,16 +51,43 @@ var backupFetchCmd = &cobra.Command{
...
+		internal.HandleBackupFetch(folder, targetBackupSelector, pgFetcher)
 	},
 }
 
+// create the BackupSelector to select the backup to fetch
+func createTargetBackupSelector(cmd *cobra.Command, args []string) (internal.BackupSelector, error) {
+	switch {
+	case len(args) == 2 && targetUserData != "":
+		fmt.Println(cmd.UsageString())
+		return nil, errors.New("Incorrect arguments. Specify backup_name OR target flag, not both.")
```
Comment (g0djan) [inline cmd/pg/backup_fetch.go:63]: Let's put all errors at the end of the switch block.

File: cmd/pg/backup_fetch.go [hunk 5]
```diff
@@ -40,16 +51,43 @@ var backupFetchCmd = &cobra.Command{
...
 
-		internal.HandleBackupFetch(folder, args[1], pgFetcher)
+		internal.HandleBackupFetch(folder, targetBackupSelector, pgFetcher)
 	},
 }
 
+// create the BackupSelector to select the backup to fetch
+func createTargetBackupSelector(cmd *cobra.Command, args []string) (internal.BackupSelector, error) {
+	switch {
+	case len(args) == 2 && targetUserData != "":
```
Comment (g0djan) [inline cmd/pg/backup_fetch.go:61]: There is no such a variable or argument as "targetUserData" in this function.

**Review by g0djan — APPROVED** [review]

**Review by g0djan — CHANGES_REQUESTED** [review]
It looks like you have 1 typo that leads to failing compilation.

File: cmd/pg/backup_fetch.go (hunk 2 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go:32]: I tried to make it clear that only one of these should be specified

Comment (usernamedt) [reply cmd/pg/backup_fetch.go:32]: maybe [backup_name | --target-user-data <data>] ?

File: cmd/pg/backup_fetch.go (hunk 3 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go:37]: I use it in createTargetBackupSelector

File: cmd/pg/backup_fetch.go (hunk 5 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go:61]: It is a global variable. It needs to be global to make it work with cobra command flag declaration.

File: cmd/pg/backup_fetch.go (hunk 1 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go]: fixed in 
9874d81

File: cmd/pg/backup_fetch.go (hunk 4 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go]: fixed in 7e364e1

File: cmd/pg/backup_fetch.go (hunk 2 above)
Comment (usernamedt) [reply cmd/pg/backup_fetch.go]: Fixed in a537a95

File: cmd/pg/backup_fetch.go (hunk 1 above)
Comment (g0djan) [reply cmd/pg/backup_fetch.go]: Probably you didn't get it. "UserData/user_data -> label" everywhere

**Review by g0djan — CHANGES_REQUESTED** [review]
Label

**Review by g0djan — APPROVED** [review]
lgtm

**Pull request merged**